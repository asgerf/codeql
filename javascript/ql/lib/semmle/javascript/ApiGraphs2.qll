import javascript
private import semmle.javascript.dataflow.internal.FlowSteps as FlowSteps
private import internal.CachedStages
private import semmle.javascript.DeepFlow
private import semmle.javascript.dataflow.internal.DataFlowNode

module API2 {
  deprecated Node root() { none() }

  /**
   * A class for contributing new steps for tracking uses of an API.
   */
  class AdditionalUseStep extends Unit {
    /**
     * Holds if use nodes should flow from `pred` to `succ`.
     */
    predicate step(DataFlow::SourceNode pred, DataFlow::SourceNode succ) { none() }
  }

  private module AdditionalUseStep {
    pragma[nomagic]
    predicate step(DataFlow::SourceNode pred, DataFlow::SourceNode succ) {
      any(AdditionalUseStep st).step(pred, succ)
    }
  }

  cached
  Node moduleImport(string name) {
    result = DataFlow::moduleImport(name)
    or
    result = DataFlow::moduleMember(name, "default")
  }

  module Node {
    Node ofType(string packageName, string typeName) {
      result.(DataFlow::SourceNode).hasUnderlyingType(packageName, typeName)
    }
  }

  class InvokeNode extends DataFlow::InvokeNode {
    pragma[inline]
    SinkNode getParameter(int i) { result = this.getArgument(i) }

    pragma[inline]
    SinkNode getAParameter() { result = this.getAnArgument() }

    pragma[inline]
    SinkNode getLastParameter() { result = this.getLastArgument() }

    pragma[inline]
    Node getReturn() { result = this }

    pragma[inline]
    Node getInstance() { result = this and this instanceof DataFlow::NewNode }
  }

  class CallNode extends InvokeNode, DataFlow::CallNode { }

  class NewNode extends InvokeNode, DataFlow::NewNode { }

  class Node extends TNode instanceof DataFlow::SourceNode {
    deprecated Node getASuccessor() { none() }

    string toString() { result = super.toString() }

    predicate hasLocationInfo(
      string filepath, int startline, int startcolumn, int endline, int endcolumn
    ) {
      super.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
    }

    pragma[inline]
    private DataFlow::SourceNode forward() {
      Deep::hasFlowTo(pragma[only_bind_out](this), pragma[only_bind_into](result))
    }

    pragma[inline]
    DataFlow::Node getAValueReachableFromSource() { result = this.forward().getALocalUse() }

    pragma[inline]
    DataFlow::SourceNode asSource() { result = this }

    pragma[inline]
    API2::CallNode getACall() { result = this.forward().getACall() }

    pragma[inline]
    API2::CallNode getMaybePromisifiedCall() {
      result = this.getACall()
      or
      result = Deep::getABoundInvocation(this, true, _)
    }

    pragma[inline]
    API2::NewNode getAnInstantiation() { result = this.forward().getAnInstantiation() }

    pragma[inline]
    API2::InvokeNode getAnInvocation() { result = this.forward().getAnInvocation() }

    pragma[inline]
    API2::Node getMember(string name) { result = Deep::getLoad(pragma[only_bind_out](this), name) }

    pragma[inline]
    API2::Node getUnknownMember() {
      DataFlow::SourceNode::Internal::dynamicPropRef(pragma[only_bind_out](this).forward(),
        pragma[only_bind_into](result)) and
      pragma[only_bind_out](result) instanceof DataFlow::PropRead
      // Note: The type check above is not strictly needed since API::Node is a subtype of SourceNode, and PropWrite is not a SourceNode.
      // TODO: Investigate which check is more efficient in practice.
    }

    pragma[inline]
    API2::Node getAMember() { result = pragma[only_bind_out](this.forward()).getAPropertyRead() }

    pragma[inline]
    API2::Node getInstance() {
      result = this.getAnInstantiation() // note: these predicates return the same thing, but with different types
    }

    pragma[inline]
    API2::SinkNode getParameter(int i) {
      Deep::argumentPassing(pragma[only_bind_out](this), i, result)
    }

    pragma[inline]
    API2::SinkNode getLastParameter() {
      result = pragma[only_bind_out](this.getACall()).getLastArgument()
      or
      result =
        TApiSyntheticCallbackArg(Deep::getABoundInvocation(pragma[only_bind_out](this), true, _))
    }

    pragma[inline]
    API2::SinkNode getAParameter() { Deep::argumentPassing(pragma[only_bind_out](this), _, result) }

    pragma[inline]
    API2::SinkNode getReceiver() {
      Deep::receiverPassing(this, result) // TODO: change to only return explicit receivers?
    }

    pragma[inline]
    API2::Node getReturn() {
      result = this.getAnInvocation() // note: these predicates return the same thing, but with different types
    }

    pragma[inline]
    API2::Node getPromised() { result = Deep::getPromised(this) }

    pragma[inline]
    API2::Node getPromisedError() { result = Deep::getPromisedError(this) }

    Node getADecoratedClass() { none() } // TODO

    Node getADecoratedMember() { none() } // TODO

    SinkNode getADecoratedParameter() { none() } // TODO

    /**
     * DEPRECATED. Use `getACall().getNumArgument()` instead, but note that each call may have a different number of arguments.
     *
     * This predicate returns the largest arity among these calls, or zero if there are no calls.
     */
    deprecated int getNumParameter() { result = max(int s | exists(this.getParameter(s))) + 1 }

    /** DEPRECATED. `API::Node` can no longer represent sinks -- use `API::SinkNode` instead. */
    deprecated DataFlow::Node asSink() { none() }

    /** DEPRECATED. `API::Node` can no longer represent sinks -- use `API::SinkNode` instead. */
    deprecated DataFlow::SourceNode getAValueReachingSink() { none() }

    /** DEPRECATED. This predicate has been renamed to `asSource`. */
    deprecated DataFlow::SourceNode getAnImmediateUse() { result = this.asSource() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachableFromSource`. */
    deprecated DataFlow::Node getAUse() { result = this.getAValueReachableFromSource() }

    /** DEPRECATED. This predicate has been renamed to `asSink` and moved to the `API::SinkNode` class. */
    deprecated DataFlow::Node getARhs() { result = this.asSink() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachingSink` and moved to the `API::SinkNode` class. */
    deprecated DataFlow::Node getAValueReachingRhs() { result = this.getAValueReachingSink() }
  }

  class BacktrackingNode extends TNode instanceof DataFlow::SourceNode {
    // TODO: include synth callback arg
    string toString() { result = this.(DataFlow::Node).toString() }

    predicate hasLocationInfo(
      string filepath, int startline, int startcolumn, int endline, int endcolumn
    ) {
      this.(DataFlow::Node).hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
    }

    pragma[inline]
    private DataFlow::SourceNode backward() {
      Deep::hasFlowTo(pragma[only_bind_into](result), pragma[only_bind_out](this))
    }

    pragma[inline]
    DataFlow::SourceNode getAValueReachingSink() { result = this.backward() }

    pragma[inline]
    DataFlow::FunctionNode getAFunction() { result = this.backward() }

    pragma[inline]
    API2::SinkNode getMember(string name) {
      // TODO: include store edges for class members
      // TODO: include reverse-load edges for stepping through long access paths
      result = this.backward().getAPropertyWrite(name).getRhs()
    }

    pragma[inline]
    API2::SinkNode getUnknownMember() {
      exists(DataFlow::PropWrite write |
        DataFlow::SourceNode::Internal::dynamicPropRef(pragma[only_bind_out](this).backward(),
          pragma[only_bind_into](write)) and
        result = write.getRhs()
      )
    }

    pragma[inline]
    API2::SinkNode getAMember() {
      // TODO: include store edges for class members
      result = this.backward().getAPropertyWrite().getRhs()
    }

    pragma[inline]
    API2::Node getInstance() { result = Deep::getClassReceiverRef(this.backward()) }

    pragma[inline]
    API2::Node getParameter(int i) { Deep::parameterDef(this.backward(), i, result) }

    pragma[inline]
    API2::Node getAParameter() { Deep::parameterDef(this.backward(), _, result) }

    pragma[inline]
    API2::Node getLastParameter() {
      // TODO: is this the best way to do this?
      exists(DataFlow::SourceNode src | src = this.backward() |
        result = src.(DataFlow::FunctionNode).getLastParameter()
        or
        result = src.(DataFlow::ClassNode).getConstructor().getLastParameter()
      )
    }

    pragma[inline]
    API2::Node getReceiver() { result = this.getAFunction().getReceiver() }

    pragma[inline]
    API2::SinkNode getReturn() { result = Deep::getPrettyReturn(this.getAFunction()) }

    pragma[inline]
    API2::SinkNode getPromised() {
      result = Deep::getStoreRhs(this.backward(), Promises::valueProp())
    }

    pragma[inline]
    API2::SinkNode getPromisedError() {
      result = Deep::getStoreRhs(this.backward(), Promises::errorProp())
    }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedClass() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedMember() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedParameter() { none() }

    /**
     * DEPRECATED. Use `getAFunction().getNumParameter()` instad, but note that different functions flowing here may
     * each have a different number of parameters.
     *
     * This predicate returns the largest number of parameters, or zero if there are no functions flowing here.
     */
    deprecated int getNumParameter() { result = max(int s | exists(this.getParameter(s))) + 1 }

    /**
     * DEPRECATED. This predicate should only be used on `API::SinkNode`. Once backtracking is performed the node is rarely
     * useful as a sink.
     *
     * Returns the receiver itself as a data-flow node, though this does not generally have the appropriate location for u
     */
    deprecated DataFlow::Node asSink() { result = this }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getAnInvocation() { none() }

    /**
     * DEPRECATED. This predicate has no result for backtracking nodes.
     * - For finding references to `this` inside a class, use `getInstance()` instead.
     * - For finding instantiation sites of a class, use a forward-tracking node instead (`API::Node`).
     */
    deprecated DataFlow::NewNode getAnInstantiation() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getACall() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getMaybePromisifiedCall() { none() }

    /** DEPRECATED. This predicate has been renamed to `asSource` and is only usable for forward-tracking nodes (`API::Node`). */
    deprecated DataFlow::SourceNode getAnImmediateUse() { none() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachableFromSource` and is only usable for forward-tracking nodes (`API::Node`). */
    deprecated DataFlow::Node getAUse() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::SourceNode asSource() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::SourceNode getAValueReachableFromSource() { none() }

    /** DEPRECATED. This predicate has been renamed to `asSink` and should only be use on the `API::SinkNode` class. */
    deprecated DataFlow::Node getARhs() { result = this.asSink() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachingSink`. */
    deprecated DataFlow::Node getAValueReachingRhs() { result = this.getAValueReachingSink() }
  }

  class SinkNode extends TDataFlowNodeOrApiNode {
    // TODO: include synth callback arg
    string toString() { result = this.(DataFlow::Node).toString() }

    predicate hasLocationInfo(
      string filepath, int startline, int startcolumn, int endline, int endcolumn
    ) {
      this.(DataFlow::Node).hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
    }

    pragma[inline]
    API2::BacktrackingNode backtrack() {
      result = pragma[only_bind_out](this).(DataFlow::Node).getALocalSource()
    }

    pragma[inline]
    DataFlow::SourceNode getAValueReachingSink() {
      result = this.backtrack().getAValueReachingSink()
    }

    pragma[inline]
    API2::SinkNode getMember(string name) { result = this.backtrack().getMember(name) }

    pragma[inline]
    API2::SinkNode getUnknownMember() { result = this.backtrack().getUnknownMember() }

    pragma[inline]
    API2::SinkNode getAMember() { result = this.backtrack().getAMember() }

    pragma[inline]
    API2::Node getInstance() { result = this.backtrack().getInstance() }

    pragma[inline]
    API2::Node getParameter(int i) {
      result = this.backtrack().getParameter(i)
      or
      Deep::promisifiedCallbackParameterDef(this, i, result)
    }

    pragma[inline]
    API2::Node getAParameter() { result = this.backtrack().getAParameter() }

    pragma[inline]
    API2::Node getLastParameter() { result = this.backtrack().getLastParameter() }

    pragma[inline]
    API2::Node getReceiver() { result = this.backtrack().getReceiver() }

    pragma[inline]
    API2::SinkNode getReturn() { result = this.backtrack().getReturn() }

    pragma[inline]
    API2::SinkNode getPromised() { result = this.backtrack().getPromised() }

    pragma[inline]
    API2::SinkNode getPromisedError() { result = this.backtrack().getPromisedError() }

    /**
     * Returns this node as a data-flow node.
     */
    pragma[inline]
    DataFlow::Node asSink() { result = this }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedClass() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedMember() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedParameter() { none() }

    /**
     * DEPRECATED. Use `getAFunction().getNumParameter()` instad, but note that different functions flowing here may
     * each have a different number of parameters.
     *
     * This predicate returns the largest number of parameters, or zero if there are no functions flowing here.
     */
    deprecated int getNumParameter() { result = max(int s | exists(this.getParameter(s))) + 1 }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getAnInvocation() { none() }

    /**
     * DEPRECATED. This predicate has no result for backtracking nodes.
     * - For finding references to `this` inside a class, use `getInstance()` instead.
     * - For finding instantiation sites of a class, use a forward-tracking node instead (`API::Node`).
     */
    deprecated DataFlow::NewNode getAnInstantiation() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getACall() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getMaybePromisifiedCall() { none() }

    /** DEPRECATED. This predicate has been renamed to `asSource` and is only usable for forward-tracking nodes (`API::Node`). */
    deprecated DataFlow::SourceNode getAnImmediateUse() { none() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachableFromSource` and is only usable for forward-tracking nodes (`API::Node`). */
    deprecated DataFlow::Node getAUse() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::SourceNode asSource() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::SourceNode getAValueReachableFromSource() { none() }

    /** DEPRECATED. This predicate has been renamed to `asSink` and should only be use on the `API::SinkNode` class. */
    deprecated DataFlow::Node getARhs() { result = this.asSink() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachingSink`. */
    deprecated DataFlow::Node getAValueReachingRhs() { result = this.getAValueReachingSink() }
  }
}
