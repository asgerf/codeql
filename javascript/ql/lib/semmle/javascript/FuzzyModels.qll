private import javascript

private DataFlow::SourceNode blockedNode() {
  exists(string name | name = result.(DataFlow::PropRead).getPropertyName() |
    name =
      any(ExternalInstanceMemberDecl decl | decl.getBaseName() = ["Object", "String", "Array"])
          .getName()
    or
    name = ["then", "catch", "finally"]
  )
  or
  exists(API::AdditionalUseStep step | step.step(_, result))
  or
  DataFlow::SharedFlowStep::step(_, result)
}

/**
 * Gets a normalized package name from `importString`.
 */
bindingset[importString]
string normalizePackageName(string importString) {
  result =
    importString
        .regexpReplaceAll("^node:", "")
        .regexpReplaceAll("\\.js$", "")
        .regexpReplaceAll("/index$", "")
}

/**
 * Gets a node believed to originate from `package`.
 *
 * The package is normalized by `normalizePackageName`.
 */
API::Node getANodeFromPackage(string package) {
  exists(string importString |
    result = API::moduleImport(importString) and
    package = normalizePackageName(importString)
  )
  or
  result = getANodeFromPackage(package).getAMember() and
  not result.asSource() = blockedNode()
  or
  result = getANodeFromPackage(package).getReturn()
  or
  result = getANodeFromPackage(package).getInstance()
  or
  result = getANodeFromPackage(package).getAParameter().getAParameter()
  or
  result = getANodeFromPackage(package).getAParameter().getAMember().getAParameter()
  or
  result = getANodeFromPackage(package).getPromised()
}

/**
 * Gets a call targeting targeting a method named `methodName` in `package`.
 */
DataFlow::InvokeNode getANamedExternalCall(string package, string methodName) {
  exists(API::Node member |
    member = getANodeFromPackage(package).getMember(methodName) and
    result = member.getAnInvocation() and
    not member.asSource() = blockedNode() and
    not (methodName = "default" and member.asSource() instanceof DataFlow::ModuleImportNode)
  )
}

/**
 * Gets a call that invokes `package` as a function (that is, its exported object is a function).
 */
DataFlow::InvokeNode getADirectExternalCall(string package) {
  result = API::moduleImport(package).getAnInvocation()
}

/**
 * Gets a call target likely targets `package`, but with no associated method name.
 *
 * For example:
 * ```javascript
 * const lib = require('example')
 *
 * let fn = lib();
 * fn() // curried calls are considered anonymous
 *
 * lib(fn => {
 *   fn(); // callback invocation is considered anonymous
 * })
 *
 * lib[complicated()](); // method call with unknown name
 * ```
 */
DataFlow::InvokeNode getAnAnonymousExternalCall(string package) {
  result = getANodeFromPackage(package).getAnInvocation() and
  not result = getANodeFromPackage(package).getMember(_).getAnInvocation() and
  not result = getADirectExternalCall(package)
}
