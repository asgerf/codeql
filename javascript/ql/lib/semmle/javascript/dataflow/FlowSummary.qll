/** Provides classes and predicates for defining flow summaries. */

private import javascript
private import semmle.javascript.dataflow.internal.sharedlib.FlowSummaryImpl as Impl
private import semmle.javascript.dataflow.internal.FlowSummaryPrivate
private import semmle.javascript.dataflow.internal.sharedlib.DataFlowImplCommon as DataFlowImplCommon
private import semmle.javascript.dataflow.internal.DataFlowPrivate

/** A callable with a flow summary, identified by a unique string. */
abstract class SummarizedCallable extends LibraryCallable, Impl::Public::SummarizedCallable {
  bindingset[this]
  SummarizedCallable() { any() }

  // TODO: rename 'propagatesFlowExt' and/or override 'propagatesFlow' directly
  /**
   * Holds if data may flow from `input` to `output` through this callable.
   *
   * `preservesValue` indicates whether this is a value-preserving step or a taint-step.
   */
  pragma[nomagic]
  predicate propagatesFlowExt(string input, string output, boolean preservesValue) { none() }

  override predicate propagatesFlow(
    string input, string output, boolean preservesValue, string model
  ) {
    this.propagatesFlowExt(input, output, preservesValue) and model = this
  }

  /**
   * Gets the synthesized parameter that results from an input specification
   * that starts with `Argument[s]` for this library callable.
   */
  DataFlow::ParameterNode getParameter(string s) {
    exists(ParameterPosition pos |
      DataFlowImplCommon::parameterNode(result, MkLibraryCallable(this), pos) and
      s = encodeParameterPosition(pos)
    )
  }
}
