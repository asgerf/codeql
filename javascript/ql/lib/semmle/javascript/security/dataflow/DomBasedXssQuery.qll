/**
 * Provides a taint-tracking configuration for reasoning about DOM-based
 * cross-site scripting vulnerabilities.
 */

import javascript
private import semmle.javascript.security.TaintedUrlSuffix
import DomBasedXssCustomizations::DomBasedXss
private import Xss::Shared as Shared
private import semmle.javascript.dataflow2.DataFlow as DataFlow2
private import semmle.javascript.dataflow2.TaintTracking as TaintTracking2
private import semmle.javascript.dataflow2.BarrierGuards
private import semmle.javascript.dataflow2.DeduplicateFlowState

/**
 * A sink that is not a URL write or a JQuery selector,
 * assumed to be a value that is interpreted as HTML.
 */
class HtmlSink extends DataFlow::Node instanceof Sink {
  HtmlSink() {
    not this instanceof WriteUrlSink and
    not this instanceof JQueryHtmlOrSelectorSink
  }
}

/** DEPRECATED: Alias for HtmlSink */
deprecated class HTMLSink = HtmlSink;

module ConfigurationArgs implements DataFlow2::StateConfigSig {
  class FlowState = DataFlow::FlowLabel;

  private predicate isSourceRaw(DataFlow::Node source, DataFlow::FlowLabel label) {
    source instanceof Source and
    (label.isTaint() or label = prefixLabel()) and
    not source = TaintedUrlSuffix::source()
    or
    source = TaintedUrlSuffix::source() and
    label = TaintedUrlSuffix::label()
  }

  private predicate isSinkRaw(DataFlow::Node sink, DataFlow::FlowLabel label) {
    sink instanceof HtmlSink and
    label = [TaintedUrlSuffix::label(), prefixLabel(), DataFlow::FlowLabel::taint()]
    or
    sink instanceof JQueryHtmlOrSelectorSink and
    label = [DataFlow::FlowLabel::taint(), prefixLabel()]
    or
    sink instanceof WriteUrlSink and
    label = prefixLabel()
  }

  import MakeDeduplicateFlowState<isSourceRaw/2, isSinkRaw/2>

  private predicate isBarrierGuard(DataFlow::BarrierGuardNode node) {
    node instanceof PrefixStringSanitizer or
    node instanceof ContainsHtmlGuard
  }

  import MakeSanitizerGuards<isBarrierGuard/1>

  predicate isBarrier(DataFlow::Node node) {
    node instanceof Sanitizer
    or
    barrierGuardBlocksNode(node)
  }

  predicate isBarrier(DataFlow::Node node, DataFlow::FlowLabel lbl) {
    barrierGuardBlocksNode(node, lbl)
    or
    // copy all taint barrier guards to the TaintedUrlSuffix/PrefixLabel label
    barrierGuardBlocksNode(node, DataFlow::FlowLabel::taint()) and
    lbl = [TaintedUrlSuffix::label(), prefixLabel()]
    or
    // any non-first string-concatenation leaf is a barrier for the prefix label.
    exists(StringOps::ConcatenationRoot root |
      node = root.getALeaf() and
      not node = root.getFirstLeaf() and
      lbl = prefixLabel()
    )
    or
    // we assume that `.join()` calls have a prefix, and thus block the prefix label.
    node = any(DataFlow::MethodCallNode call | call.getMethodName() = "join") and
    lbl = prefixLabel()
    or
    isOptionallySanitizedNode(node) and
    lbl = [DataFlow::FlowLabel::taint(), prefixLabel(), TaintedUrlSuffix::label()]
    or
    TaintedUrlSuffix::isBarrier(node, lbl)
    or
    deduplicationBarrier(node, lbl)
  }

  predicate isAdditionalFlowStep(
    DataFlow::Node node1, FlowState state1, DataFlow::Node node2, FlowState state2
  ) {
    TaintedUrlSuffix::step(node1, node2, state1, state2)
    or
    exists(DataFlow::Node operator |
      StringConcatenation::taintStep(node1, node2, operator, _) and
      StringConcatenation::getOperand(operator, 0).getStringValue() = "<" + any(string s) and
      state1 = TaintedUrlSuffix::label() and
      state2.isTaint()
    )
    or
    // steps out of taintedSuffixlabel to taint-label are also steps to prefixLabel.
    TaintedUrlSuffix::step(node1, node2, TaintedUrlSuffix::label(), DataFlow::FlowLabel::taint()) and
    state1 = TaintedUrlSuffix::label() and
    state2 = prefixLabel()
    or
    exists(DataFlow::FunctionNode callback, DataFlow::Node arg |
      any(JQuery::MethodCall c).interpretsArgumentAsHtml(arg) and
      callback = arg.getABoundFunctionValue(_) and
      node1 = callback.getReturnNode() and
      node2 = callback and
      state1 = state2
    )
    or
    deduplicationStep(node1, state1, node2, state2)
  }
}

module Configuration = TaintTracking2::GlobalWithState<ConfigurationArgs>;

/**
 * A taint-tracking configuration for reasoning about XSS.
 * Both ordinary HTML sinks, URL sinks, and JQuery selector based sinks.
 * - HTML sinks are sinks for any tainted value
 * - URL sinks are only sinks when the scheme is user controlled
 * - JQuery selector sinks are sinks when the tainted value can start with `<`.
 *
 * The above is achieved using three flow labels:
 * - TaintedUrlSuffix: a URL where the attacker only controls a suffix.
 * - Taint: a tainted value where the attacker controls part of the value.
 * - PrefixLabel: a tainted value where the attacker controls the prefix
 */
deprecated class Configuration extends TaintTracking::Configuration {
  Configuration() { this = "HtmlInjection" }

  override predicate isSource(DataFlow::Node source, DataFlow::FlowLabel label) {
    ConfigurationArgs::isSource(source, label)
  }

  override predicate isSink(DataFlow::Node sink, DataFlow::FlowLabel label) {
    ConfigurationArgs::isSink(sink, label)
  }

  override predicate isSanitizer(DataFlow::Node node) { ConfigurationArgs::isBarrier(node) }

  override predicate isLabeledBarrier(DataFlow::Node node, DataFlow::FlowLabel lbl) {
    ConfigurationArgs::isBarrier(node, lbl)
  }

  override predicate isAdditionalFlowStep(
    DataFlow::Node node1, DataFlow::Node node2, DataFlow::FlowLabel state1,
    DataFlow::FlowLabel state2
  ) {
    ConfigurationArgs::isAdditionalFlowStep(node1, state1, node2, state2)
    or
    // inherit all ordinary taint steps for the prefix label
    state1 = prefixLabel() and
    state2 = prefixLabel() and
    TaintTracking::sharedTaintStep(node1, node2)
    or
    // inherit some ordinary taint steps for tainted-url-suffix label
    TaintedUrlSuffix::preservingTaintStep(node1, node2, state1, state2)
  }
}

private class PrefixStringSanitizerActivated extends TaintTracking::SanitizerGuardNode,
  PrefixStringSanitizer
{
  PrefixStringSanitizerActivated() { this = this }
}

private class PrefixStringActivated extends DataFlow::FlowLabel, PrefixString {
  PrefixStringActivated() { this = this }
}

private class QuoteGuard extends TaintTracking::SanitizerGuardNode, Shared::QuoteGuard {
  QuoteGuard() { this = this }
}

private class ContainsHtmlGuard extends TaintTracking::SanitizerGuardNode, Shared::ContainsHtmlGuard
{
  ContainsHtmlGuard() { this = this }
}
