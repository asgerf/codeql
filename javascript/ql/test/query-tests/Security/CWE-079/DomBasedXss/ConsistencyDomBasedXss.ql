import javascript
import testUtilities.ConsistencyChecking
import semmle.javascript.security.dataflow.DomBasedXssQuery

class ConsistencyConfig extends ConsistencyConfiguration {
  ConsistencyConfig() { this = "ConsistencyConfig" }

  override DataFlow::Node getAnAlert() { Configuration::flow(_, result) }
}
