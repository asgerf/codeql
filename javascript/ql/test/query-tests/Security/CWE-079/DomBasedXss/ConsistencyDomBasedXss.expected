| query-tests/Security/CWE-079/DomBasedXss/tst.js:189 | expected an alert, but found none | NOT OK |  |
| query-tests/Security/CWE-079/DomBasedXss/tst.js:191 | expected an alert, but found none | NOT OK |  |
| query-tests/Security/CWE-079/DomBasedXss/tst.js:192 | expected an alert, but found none | NOT OK |  |
| query-tests/Security/CWE-079/DomBasedXss/tst.js:193 | expected an alert, but found none | NOT OK |  |
| query-tests/Security/CWE-079/DomBasedXss/tst.js:288 | expected an alert, but found none | NOT OK |  |
| query-tests/Security/CWE-079/DomBasedXss/tst.js:360 | expected an alert, but found none | NOT OK. (this is a DOM-node); |  |
| query-tests/Security/CWE-079/DomBasedXss/tst.js:363 | expected an alert, but found none | NOT OK. |  |
| query-tests/Security/CWE-079/DomBasedXss/tst.js:465 | expected an alert, but found none | NOT OK |  |
| query-tests/Security/CWE-079/DomBasedXss/tst.js:467 | expected an alert, but found none | NOT OK |  |
